src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1081,6 +1082,73 @@ private void ApplyTextDocumentChange(DocumentId documentId, SourceText newText)
             }
         }
 
+        protected override void ApplyDocumentInfoChanged(DocumentId id, DocumentInfo updatedInfo)
+        {
+            var document = CurrentSolution.GetDocument(id);
+
+            FailIfDocumentInfoChangesNotSupported(document, updatedInfo);
+
+            if (document.Name != updatedInfo.Name)
+            {
+                GetProjectData(updatedInfo.Id.ProjectId, out var _, out var project);
+                var projectItemForDocument = project.FindItem(document.Name, StringComparer.InvariantCultureIgnoreCase); //?
+
+                if (projectItemForDocument == null)
+                {
+                    // TODO: The document might actually belong to a Shared Project.
+                    return;
+                }
+
+                // Must save the document first for things like Breakpoints to be preserved.
+                projectItemForDocument.Save();
+
+                var uniqueName = projectItemForDocument.Collection.GetUniqueName(
+                    Path.GetFileNameWithoutExtension(updatedInfo.Name),
+                    Path.GetExtension(updatedInfo.Name));
+
+                // By setting this property, Visual Studio will perform the file rename, which 
+                // will cause the workspace's current solution to update and will fire the 
+                // necessary workspace changed events.
+                projectItemForDocument.Name = uniqueName;
+            }
+        }
+
+        /// <summary>
+        /// The VisualStudioWorkspace currently supports only a subset of DocumentInfo changes.
+        /// </summary>
+        private void FailIfDocumentInfoChangesNotSupported(CodeAnalysis.Document document, DocumentInfo updatedInfo)
+        {
+            if (document.SourceCodeKind != updatedInfo.SourceCodeKind)
+            {
+                throw new InvalidOperationException(
+                    $"Unexpected DocumentInfo change to {nameof(document.SourceCodeKind)}.");
+            }
+
+            if (document.FilePath != updatedInfo.FilePath)
+            {
+                throw new InvalidOperationException(
+                    $"Unexpected DocumentInfo change to {nameof(document.FilePath)}.");
```

Comment:
This seems weird... so the name would change, but teh path would have to be the same...?  that seems  odd to me.

Reply:
Yeah... That was the driving scenario for Rename File to Match Type.

Reply:
i would prefer that the message be more like "This workspace does not support changing the filepath of a doucment"

Reply:
Done.